File: src/profile/games/mo6tw/tipsmenu.cpp
#include "tipsmenu.h"
#include "../../profile_internal.h"
#include "../../../renderer/renderer.h"
#include "../../../log.h"
#include "../../../text/text.h"

#include "../../ui/tipsmenu.h"
#include "../../../game.h"
#include "../../../ui/ui.h"
#include "../../../games/mo6tw/tipsmenu.h"

namespace Impacto {
namespace Profile {
namespace MO6TW {
namespace TipsMenu {

void Configure() {
  GetMemberArray<Sprite>(TipThumbnails, 37, "Thumbnails");
  TipTextOnlyThumbnail = EnsureGetMember<Sprite>("TextOnlyThumbnail");
  ThumbnailPosition = EnsureGetMember<glm::vec2>("ThumbnailPosition");
  auto str = EnsureGetMember<char const*>("CategoryString");
  TextGetSc3String(str, CategoryString);

  DefaultColorIndex = EnsureGetMember<int>("DefaultColorIndex");
  UnreadColorIndex = EnsureGetMember<int>("UnreadColorIndex");
  NameInitialBounds = EnsureGetMember<RectF>("NameInitialBounds");
  NameFontSize = EnsureGetMember<float>("NameFontSize");
  PronounciationInitialBounds =
      EnsureGetMember<RectF>("PronounciationInitialBounds");
  PronounciationFontSize = EnsureGetMember<float>("PronounciationFontSize");
  CategoryInitialBounds = EnsureGetMember<RectF>("CategoryInitialBounds");
  CategoryEndX = EnsureGetMember<float>("CategoryEndX");
  CategoryFontSize = EnsureGetMember<float>("CategoryFontSize");
  SortStringTable = EnsureGetMember<int>("SortStringTable");
  SortStringIndex = EnsureGetMember<int>("SortStringIndex");
  TipListInitialY = EnsureGetMember<float>("TipListInitialY");
  TipListCategoriesPerPage = EnsureGetMember<int>("TipListCategoriesPerPage");
  TipListMaxPages = EnsureGetMember<int>("TipListMaxPages");
  TipListEntryBounds = EnsureGetMember<RectF>("TipListEntryBounds");
  TipListEntryFontSize = EnsureGetMember<float>("TipListEntryFontSize");
  TipListYPadding = EnsureGetMember<float>("TipListYPadding");
  TipListEntryHighlightOffset =
      EnsureGetMember<glm::vec2>("TipListEntryHighlightOffset");
  TipListEntryNameXOffset = EnsureGetMember<float>("TipListEntryNameXOffset");
  TipListEntryNewText = EnsureGetMember<std::string>("TipListEntryNewText");
  TipListEntryNewOffset = EnsureGetMember<float>("TipListEntryNewOffset");
  TipListEntryLockedTable = EnsureGetMember<int>("TipListEntryLockedTable");
  TipListEntryLockedIndex = EnsureGetMember<int>("TipListEntryLockedIndex");
  NumberLabelStrTable = EnsureGetMember<int>("NumberLabelStrTable");
  NumberLabelStrIndex = EnsureGetMember<int>("NumberLabelStrIndex");
  NumberLabelPosition = EnsureGetMember<glm::vec2>("NumberLabelPosition");
  NumberLabelFontSize = EnsureGetMember<float>("NumberLabelFontSize");
  NumberBounds = EnsureGetMember<RectF>("NumberBounds");
  NumberFontSize = EnsureGetMember<float>("NumberFontSize");
  PageSeparatorTable = EnsureGetMember<int>("PageSeparatorTable");
  PageSeparatorIndex = EnsureGetMember<int>("PageSeparatorIndex");
  PageSeparatorPosition = EnsureGetMember<glm::vec2>("PageSeparatorPosition");
  PageSeparatorFontSize = EnsureGetMember<float>("PageSeparatorFontSize");
  CurrentPageBounds = EnsureGetMember<RectF>("CurrentPageBounds");
  TotalPagesBounds = EnsureGetMember<RectF>("TotalPagesBounds");

  auto drawType = Game::DrawComponentType::_from_integral_unchecked(
      EnsureGetMember<uint8_t>("DrawType"));

  UI::TipsMenuPtr = new UI::MO6TW::TipsMenu();
  UI::Menus[drawType].push_back(UI::TipsMenuPtr);
}

}  // namespace TipsMenu
}  // namespace MO6TW
}  // namespace Profile
}  // namespace Impacto

File: profiles/mo6tw/hud/tipsmenu.js
root.TipsMenu = {
    Type: TipsMenuType.MO6TW,
    BackgroundSprite: "TipsMenuBackground",
    Thumbnails: [],
    TextOnlyThumbnail: "TipsTextOnlyThumbnail",
    ThumbnailPosition: { X: 413, Y: 45 },
    CategoryString: "【 】",
    DefaultColorIndex: 0,
    UnreadColorIndex: 70,
    NameInitialBounds: { X: 703, Y: 117, Width: 0, Height: 0 },
    NameFontSize: 32,
    PronounciationInitialBounds: { X: 703, Y: 87, Width: 0, Height: 0 },
    PronounciationFontSize: 20,
    CategoryInitialBounds: { X: 1135, Y: 206, Width: 0, Height: 0 },
    CategoryEndX: 1200,
    CategoryFontSize: 32,
    SortStringTable: 2,
    SortStringIndex: 5,
    TipListInitialY: 87,
    TipListCategoriesPerPage: 5,
    TipListMaxPages: 10,
    TipListEntryBounds: { X: 69, Y: 87, Width: 300, Height: 24 },
    TipListEntryFontSize: 20,
    TipListYPadding: 20,
    TipListEntryHighlightOffset: { X: 0, Y: -9 },
    TipListEntryNameXOffset: 50,
    TipListEntryNewText: "New",
    TipListEntryNewOffset: 261,
    TipListEntryLockedTable: 2,
    TipListEntryLockedIndex: 7,
    NumberLabelStrTable: 2,
    NumberLabelStrIndex: 8,
    NumberLabelPosition: { X: 1070, Y: 45 },
    NumberLabelFontSize: 32,
    NumberBounds: { X: 1130, Y: 45, Width: 0, Height: 0 },
    NumberFontSize: 32,
    PageSeparatorTable: 2,
    PageSeparatorIndex: 9,
    PageSeparatorPosition: { X: 1154, Y: 645 },
    PageSeparatorFontSize: 32,
    CurrentPageBounds: { X: 1130, Y: 645, Width: 0, Height: 0 },
    TotalPagesBounds: { X: 1184, Y: 645, Width: 0, Height: 0 },
    FadeInDuration: 0.2,
    FadeOutDuration: 0.2
};

root.Sprites["TipsMenuBackground"] = {
    Sheet: "Tips",
    Bounds: { X: 0, Y: 0, Width: 1280, Height: 720 }
};

root.Sprites["TipsTextOnlyThumbnail"] = {
    Sheet: "Tips",
    Bounds: { X: 1300, Y: 0, Width: 256, Height: 192 }
};

var firstX = 0;
var firstY = 0;

for (var i = 1; i < 38; i++) {
    root.Sprites["TipThumbnail" + i] = {
        Sheet: "TipThumbnails",
        Bounds: {
            X: firstX,
            Y: firstY,
            Width: 256,
            Height: 192
        }
    };
    root.TipsMenu.Thumbnails.push("TipThumbnail" + i);

    if (i % 8 == 0) {
        firstX = 0;
        firstY += 192;
    } else {
        firstX += 256;
    }
}
